README.md
 ## Electsys++

[![Chrome Web Store](https://img.shields.io/chrome-web-store/d/ilmpacenmcgknoogkhpigakpoocpjmpl.svg)](https://chrome.google.com/webstore/detail/electsys%20%20-%E4%BA%A4%E5%A4%A7%E9%80%89%E8%AF%BE%E6%8F%92%E4%BB%B6/ilmpacenmcgknoogkhpigakpoocpjmpl)

Electsys++ SJTU electsys optimizer.
上海交通大学本科教学信息服务网优化插件。

本着交流学习、方便同学的目的将此插件建设为**开源项目**！本项目遵循 GPL v3 协议，欢迎各位同学参与贡献！您可以提交[问题反馈](https://github.com/laohyx/electsys/issues)，也可以提交 [Pull Request](https://github.com/laohyx/electsys/pulls) 帮助完善代码。


### 安装

#### 通过 Chrome Webstore（推荐）

请访问 Chrome Webstore（被墙）安装插件：

https://chrome.google.com/webstore/detail/electsys%20%20-%E4%BA%A4%E5%A4%A7%E9%80%89%E8%AF%BE%E6%8F%92%E4%BB%B6/ilmpacenmcgknoogkhpigakpoocpjmpl

#### 通过 crx 文件

1. 在 [Release](https://github.com/laohyx/electsys/releases) 页面下载最新的 crx 文件
2. 点击右上角 __菜单->更多工具->扩展程序__
3. 将 crx 文件从资源管理器中 __拖入“扩展程序”窗口__ 即可
4. 如安装失败，请确认 (1) 操作是否正确 (2) crx 文件是否来自 [Release](https://github.com/laohyx/electsys/releases) 页面或网上应用店

### 其他浏览器上的插件

#### Safari
https://github.com/dyweb/electsys-safari

#### Firefox
https://github.com/hczhcz/electsys

#### Edge

1. 在 [Release](https://github.com/laohyx/electsys/releases) 页面下载最新版本的 __Source Code__ 压缩包并在本地解压缩
2. 在 https://www.microsoft.com/en-us/store/p/microsoft-edge-extension-toolkit/9nblggh4txvb 安装 __Microsoft Edge Extension Toolkit__
3. 打开 Toolkit ,点击 __Load extension to convert__ ,选择刚才解压的文件夹载入，等待载入完成(完成后页面中间会有提示 __Your extension is ready for testing__)
4. 打开 Edge ,在地址栏输入并访问 `about:flags`
5. 勾选 __“启用开发人员扩展功能(这可能让设备处于危险之中) ”__
6. 点击右上角 __菜单(...)->扩展->加载扩展__
7. 选择插件所在的文件夹即可载入

(由于插件没有上线到应用商店，浏览器启动时会提示已关闭来自未知源的扩展;可以从 __菜单(...)->扩展__ 进行启用)

### 更新日志

请参阅 [CHANGELOG](CHANGELOG) 文件。

### 开源许可

**GPL v3** 或更新的版本。详情请参阅 [LICENSE](LICENSE) 文件。


CHANGELOG
更新日志（从 1.9.5 才开始写）：

2.6.2 - 2019/6/15
    提升验证码识别流程的稳定性 (#39)

2.6.1 - 2019/3/17
    修复验证码功能无法关闭的问题
    修复部分情况拖动评教页面滚动条数值无响应的问题
    修复了新版 jAccount 页面验证码无法识别的问题 (#38)
    更新项目许可证为 GPL v3

2.6 - 2017/4/10
    增加 jAccount 自动登录 (#26)
    修复夏季学期选课小课表无法显示的问题 (#27)

2.5.4 - 2017/1/17
    暂时关闭首页成绩查询

2.5.3 - 2016/12/25
    更新通识课列表

2.5.2 - 2016/12/13
    修复小课表无法折叠和功能开关失效的问题
    修复从推荐课表选课时点击教师时会自动跳转的问题 (#24)
    修复在推荐课表页面会出现推荐课表的小课表的问题

2.5.1 - 2016/12/12
    修复评教页面滑条失效的问题 (#17)

2.5 - 2016/11/3
    升级至 jQuery 3.1
    增加插件配置页面
    恢复首页成绩查询
    优化成绩和绩点查询筛选
    优化代码架构
    修复成绩和绩点查询中等第制成绩无法计算的问题 (#14)

2.4.2 - 2016/9/18
    增加核心课程成绩页面的 GPA 计算
    更新通识课列表
    在选课页面查看课程安排时可优先使用点击“查”按钮时时缓存的页面
    修复修业查看中正在修读课程列入 GPA 计算的问题 (#11)

2.4.1 - 2016/9/4
    去除了广告。发布到了应用商店。

2.4 - 2014/1/27
        tq5124, 更新日志改写在git中
    在”修业查看“页面中增加所有课程的 GPA 和各小项的 GPA 的显示
    修复小学期选课页面通识课没有高亮标注的问题
    更新通识课列表
    优化选课前的注意事项页面
    从 2.4 开始将只支持 Chrome 49+ 版本，请及时升级浏览器

2.3.19 - 2013/12/11
    优化处理队列，加大间隔，增加错误重试

2.3.18 - 2013/12/11
    选课网限制刷新频率，修改为手动点查

2.3.14 - 2012/12/26
    修改“刷新信息”按钮的样式。
    计划下一版本开源啊！
    
2.3.13 - 2012/11/12
    （为什么空下那么多版本号？)
    修复查积点的小bug。（招安= =他们还懒得搞呢）
    
    
2.3.8/2.2.9 - 2012/06/28
    关闭左边菜单栏优化。（唉，被招安了。。）
    
2.3.7 - 2012/06/23
    修改版本号显示方式，同步与插件版本更新。
    
2.3.6 - 2012/06/22
    修改查任选课中的一个经典BUG（竟然将POST data中的‘$’文本替换成了jQuery...）
    增加对于无安排的课程出现“无”的提示。
    姑娘21号晚回家了。祝她有个美好的周末~
    
2.3.5 - 2012/06/21
    美化一下logo。
    增加了删除老插件的功能，成功避免冲突。

2.3.3 - 2012/06/20
    修改logo。
    修复“参数错误的bug”，在读取满不满时点击查询，不会出现“参数错误”的提示。
    
2.3.1 - 2012/06/18 (WebStore分支，难道开创了奇偶数版本的时代？)
    取2.1.9的查课程代码，并加以修改，更快速稳定。
    （还是靠学校开放接口啊…………）
    
2.2.6 - 2012/06/16
    修复首页成绩显示的BUG。

2.2.5 - 2012/06/11
    不做测试悔死人啊！！
    官网链接竟然不分大小写的- -

    2.2.4 - 2012/06/11
    不做测试悔死人啊！！
    修正一小小bug。。。
2.2.3 - 2012/06/11
    首页出现本学期的成绩。

2.2.2 - 2012/06/05
    NHCE 第三册、第四册的听力部分ok
    甚至修改了Next按键的链接，直接跳过说的部分。
    
    2.2.1 - 2012/06/04
    划时代地加入了NCHE（新视野大学英语）网站的答题功能。
    第三册的听力部分全部自动答题，s和n键可以用来用为Submit和Next.
    在纠结用不用把第二册和第四册做掉呢。。。。。

2.2.0 - 2012/05/25
    2012-2013第一学期选课开始
    他们关闭了查课，那么我就用我自己的网站来查！
    修改lesson.js里post函数链接，直接从electsys.net得到。

2.1.17 - 2012/05/09
    暑假小学期第一次更新，功能不多，把“满不满”样式恢复到传统模式。。。
    求JWC给我点接口什么的……
    
2.1.14 - 2012/04/01
    愚人节快乐！
    
2.1.10 - 2012/02/12
    选课网增加对POST频率的检测机制，暂时关闭POST功能。
    尼玛难道要做不下去了么？
    
2.1.9 - 2012/01/??
    不知道做了些什么。。
    
2.1.8 - 2011/12/21
    那个问题果然没解决，手动去掉关闭功能。
    
2.1.7 - 2011/12/19
    修改控制页面的链接，解决缓存问题（也不知能不能解决）

2.1.5 - 2011/12/10
    根据jwc通知增加冲抵通识课的任选课列表。
    增强域名兼容性。
    
2.1.4 - 2011/12/8 
    修改部分代码改为jQuery API，增强兼容性。
    增加lesson_enable_check功能，可以选择开关功能，避免抢选时增加服务器压力。
    
2.1.3 - 2011/12/6 早晨
    修改颜色块的样式，好看多了吧……
    修改统计链接的增加方式，减少性能消耗。

2.1.2 - 2011/12/5 晚上
    修改draw div的位置，使得其绘制不影响表格各列的宽度。

2.1.1 - 2011/12/5 晚上
    修复bug：electsys0下post数据到electsys.sjtu.edu.cn   增加base_url解决
    修改推广链接，到electsys.net，总比分享到人人好吧……
    
2.1.0 - 2011/12/5 下午
    增加了好多功能啊。。
    修改选课时标记颜色，比原来的柔和多了……
    点课程的单选圆圈就可以查看课程安排了！
    在课程安排中显示该老师的评教数据！
    
2.0.5 - 2011/12/1 晚上
    修正2.0.4增加的联动功能没用的bug - -#
    Bug reported by Popacai.(专业测试员~)
    
2.0.4 - 2011/12/1 下午
    增加分数联动功能，但是分数太整齐...

2.0.3 - 2011/12/1
    优化判断链接，支持electsys0 和不带0的
    增加“快速评教功能”
    纪念下，2011年11月29日，electsys.net上线

2.0.2 - 2011/11/22
    将判断链接由electsys0.sjtu.edu.cn/*改为electsys.sjtu.edu.cn/*
    
2.0.1 - 2011/10/22       (from 2011/9/10)
    修改后台为异步post方式。
    在登录后停止后台查询。
    //在首页支持记录账号数据。(关闭)

2.0.0 - 2011/8/3 （未发布）
    增加后台查成绩功能。
    增加工具栏图标，有各种小提示。。。
    
1.9.5 - 2011/7/18
    将dom操作全部改为jquery API，增加可移植性。
    将load.js分解成四个文件，方便管理。
    修复算平均分时遇到字母时出错的bug。
    增加算平均分全选的功能。


content_script/page/nhce.js
/**
 * Electsys++ Project
 * ----------------------------
 * nhce 辅助模块
 * For 苏文. 2012/06/04
 */

function nhce_main(){
	if (!inUrl("202.120.59.230"))
        return 0;
		

	answer03 = {"1": {"1": ["C", "D", "A", "D", "B", "A", "D", "C", "A", "B"], "2": ["A", "C", "B", "B", "D"], "3": ["B", "A", "C", "C", "C"], "4": ["a journalism degree", "pretty good", "about what she has", "you just want more", "protective of his family"], "5": ["C", "C", "A", "D", "B"], "6": ["A", "C", "D", "C", "D"], "7": ["slice", "misunderstandings", "beautiful", "benefits", "wellness", "range", "explicit", "has been tracking more than a million subjects since 1979", "have fewer heart attacks and lower cancer rates", "a strong sense of connection to others and in satisfying relationship"]}, "2": {"1": ["C", "C", "B", "A", "D", "B", "C", "D", "A", "A"], "2": ["D", "C", "A", "B", "A"], "3": ["B", "D", "D", "C", "A"], "4": ["the quality of school menus", "moving into the neighbourhood", "eats fast food", "the Chicago study", "children's eating habits"], "5": ["A", "B", "C", "D", "B"], "6": ["A", "B", "D", "C", "D"], "7": ["advancements", "physical", "depression", "condition", "mentioned", "admit", "combat", "high-fiber and vitamin-rich foods, such as vegetables and fruits", "should give up smoking if they haven't already done so", "No section of the population can benefit from exercise."]}, "3": {"1": ["D", "C", "B", "C", "C", "B", "A", "C", "B", "A"], "2": ["B", "C", "A", "B", "D"], "3": ["A", "B", "A", "C", "A"], "4": ["is studying for a doctorate in religion", "is marked by chance encounters", "only knew a little Italian", "was attacked by an intruder in her home", "began to live a life with vivid moments"], "5": ["C", "D", "D", "A", "B"], "6": ["B", "B", "C", "D", "B"], "7": ["experience", "indicate", "distinguish", "Successful", "long-term", "achievements", "follow", "unsuccessful people just let life happen by accident", "and they aren't difficult for people to attain", "what to go after and in what direction to aim your life"]}, "4": {"1": ["B", "B", "C", "D", "A", "B", "D", "C", "D", "A"], "2": ["C", "D", "B", "B", "A"], "3": ["B", "A", "D", "D", "B"], "4": ["for decades", "healthy and active lifestyle", "meet consumer demands","spur competitiveness"], "5": ["D", "C", "C", "B", "B"], "6": ["C", "A", "B", "D", "B"], "7": ["studios", "counterparts", "emphasize", "innovation", "specialized", "energized", "filmmakers", "They began playing with and contradicting the conversations of Hollywood", "Furthermore, their considerable financial success and crossover into popular culture", "Some people have taken advantage of this rise in popularity"]}, "5": {"1": ["B", "C", "D", "A", "B", "B", "B", "A", "D", "D"], "2": ["B", "C", "D", "D", "A"], "3": ["A", "C", "B", "C", "C"], "4": ["do it at pains", "I love you", "their appearance", "you don't want to answer", "think of her in your heart"], "5": ["D", "A", "A", "B", "C"], "6": ["B", "D", "C", "A", "A"], "7": ["Yet", "elaborated", "outward", "kneel", "embarrassed", "express", "gathered", "but over time, I started to enjoy them more and more", "Now that I've talked to you, I feel ten years younger.", "people shouldn't live their lives always thinking about the past."]}, "6": {"1": ["D", "D", "D", "B", "C", "A", "C", "C", "A", "C"], "2": ["A", "C", "B", "B", "D"], "3": ["C", "D", "C", "D", "A"], "4": ["Islamabad", "a tall Islamabad building", "the Kashmirian Mountains", "a heavily militarized region", "India"], "5": ["A", "B", "C", "D", "A"], "6": ["A", "A", "B", "D", "A"], "7": ["thickness", "frightened", "swallowed", "disappeared", "floods", "revisit", "landslide", "keep us away from mountains the rest of our lives", "something like the Yellowstone earthquake does not happen very often", "few of us will suffer because of such a disaster."]}, "7": {"1": ["C", "A", "C", "B", "D", "A", "B", "B", "A", "C"], "2": ["A", "B", "B", "C", "D"], "3": ["A", "A", "D", "C", "B"], "4": ["The Day to Day", "To try to go after the online music business", "Apple", "It has been negotiating licenses for songs", "Gate's leaving will benefit it"], "5": ["D", "C", "D", "D", "A"], "6": ["D", "C", "B", "C", "A"], "7": ["industry", "option", "preferable", "starve", "blame", "occupations", "force", "But if we admit tradition more than we should", "insist on trying to plot the future by the past", "to accept the help which tradition can bring"]}, "8": {"1": ["B", "D", "C", "A", "A", "A", "C", "C", "A", "B"], "2": ["A", "A", "D", "D", "C"], "3": ["D", "C", "A", "B", "C"], "4": ["human beings", "maternal twins", "genes that cause disease", "manny disease", "an environmental component"], "5": ["B", "C", "B", "C", "D"], "6": ["C", "A", "B", "C", "C"], "7": ["ethics", "prediction", "technique", "adapted", "artificial", "urgent", "inconsistencies", "\"Genetic engineering is very exciting and important technology\"", "but you have to ask fundamental questions before it is too late", "the way the technology is racing far ahead of public understanding"]}, "9": {"1": ["B", "D", "A", "C", "A", "B", "C", "C", "D", "D"], "2": ["A", "B", "D", "C", "D"], "3": ["D", "D", "A", "C", "A"], "4": ["her benefits were fabulous", "really, really tough", "got a part-time job", "on the Internet/online", "lazy"], "5": ["C", "A", "B", "B", "C"], "6": ["D", "C", "C", "C", "B"], "7": ["university", "expect", "marriage", "mature", "exaggerating", "Undergraduates", "candidates", "The answer to whether or not marriage in school should be allowed.", "As for those who are attending universities at an older age.", "the marriage of some undergraduates will inevitably influence  other students"]}, "10": {"1": ["A", "B", "B", "C", "D", "A", "B", "D", "C", "D"], "2": ["B", "D", "C", "A", "C"], "3": ["A", "C", "B", "D", "A"], "4": ["in 1776", "They worked together for the same goal", "Fifty years", "No ,he was ailing", "Yes, they did."], "5": ["C", "A", "B", "D", "B"], "6": ["D", "A", "A", "D", "B"], "7": ["organisms", "untreated", "occurences", "agriculture", "encouraged", "administrations", "attributed", "Some things have been done to address this problem", "in an effort to raise the public's awareness of environmental protection", "to punish enterprises that drain untreated wastes into the rivers and oceans"]}};
	
	answer04 = {"1": {"1": ["A", "C", "B", "B", "D", "B", "A", "C", "C", "D"], "2": ["A", "B", "A", "C", "B"], "3": ["A", "A", "D", "B", "C"], "4": ["married in St. Paul's Cathedral", "was a guest at the wedding ceremony", "wore a hat with flowers at the wedding", "is remembered as having been naughty", "was one of the designers of the wedding dress"], "5": ["D", "A", "B", "B", "A"], "6": ["C", "A", "B", "B", "A"], "7": ["shortage", "assigned", "centered", "hospitalization", "treatment", "colleague", "decentralized", "There are nurse-managers instead of head-nurses", "decide among themselves who will work what shifts and when", "an equal with other vice presidents of the hospital"]}, "2": {"1": ["D", "D", "D", "C", "B", "A", "A", "B", "C", "D"], "2": ["D", "A", "B", "B", "A"], "3": ["B", "C", "D", "A", "C"], "4": ["C", "A", "A", "C", "D"], "5": ["D", "A", "B", "B", "A"], "6": ["B", "D", "A", "C", "A"], "7": ["sketch", "shadow", "paintings", "abroad", "Europe", "jewels", "exquisite", "pictures of rooms with handsomely dressed people in them", "not just their clothes or the lines of their faces", "but he was far greater than they would ever become"]}, "3": {"1": ["C", "D", "C", "B", "B", "B", "A", "C", "A", "B"], "2": ["A", "B", "D", "C", "B"], "3": ["A", "B", "D", "B", "C"], "4": ["An Australian scientist who won the Nobel Prize.", "The mysterious field of infectious diseases.", "By accident.", "It was probably extremely significant.", "He couldn't handle all that."], "5": ["B", "C", "A", "D", "C"], "6": ["B", "A", "C", "B", "C"], "7": ["September", "retire", "retirement", "reduction", "practical", "pensions", "leisure", "The club arranges discussion groups and handicraft sessions", "a member can attend any course held there free of charge", "the financial section on Mondays and Wednesdays between six and eight p.m."]}, "4": {"1": ["B", "A", "D", "D", "A", "D", "C", "C", "A", "A"], "2": ["C", "B", "B", "B", "C"], "3": ["A", "C", "B", "A", "C"], "4": ["became Bill Gates' greatest contribution", "makes use of Gates' system", "was the plaything of nerds", "became a business tool", "made it a wish to dominate like Bill Gates", "was not fit to comment on upcoming innovation"], "5": ["B", "A", "B", "C", "A"], "6": ["A", "A", "B", "B", "A"], "7": ["fundamental", "dramatically", "majority", "workplace", "self-employed", "breadth", "notions", "its applications in personal computers, digital communications, and factory robots", "still unimagined technology could produce a similar wave of dramatic changes", "will have the greatest advantage and produce the most wealth"]}, "5": {"1": ["A", "D", "D", "D", "B", "C", "D", "C", "A", "D"], "2": ["B", "B", "A", "C", "C"], "3": ["D", "A", "B", "B", "A"], "4": ["how they are turning men off", "an idea in one's head about what the perfect man looks like", "a much younger version of yourself", "more advice on dating and relationships", "burning the perfect man checklist"], "5": ["C", "D", "A", "B", "D"], "6": ["A", "C", "A", "C", "D"], "7": ["emerging", "residents", "participants", "companionship", "soldiers", "isolated", "extraordinary", "who happen to live by themselves die at twice the rate of those live with others", "It's clear that reaching out to others can help our bodies thrive", "only 5 percent of U.S. households consisted of one person living alone"]}, "6": {"1": ["C", "B", "C", "D", "B", "C", "A", "D", "D", "B"], "2": ["A", "B", "C", "C", "D"], "3": ["C", "D", "D", "A", "B"], "4": ["Centennial Olympic Park", "North Cardwell, New Jersey", "Oklahoma City, Oklahoma", "the FBI laboratories", "ground zero", "an Atlanta abortion clinic"], "5": ["D", "C", "B", "A", "A"], "6": ["C", "B", "C", "A", "D"], "7": ["impeach", "scandal", "gambling", "Representatives", "accusations", "procedures", "opposition", "resigned as Secretary of Social Welfare and urged the President to resign", "five economic advisers to the President have resigned", "the Judicial Committee and Economic Affairs Committee in the House of Representatives"]}, "7": {"1": ["B", "C", "A", "A", "D", "C", "B", "D", "A", "B"], "2": ["D", "A", "C", "C", "B"], "3": ["C", "C", "D", "B", "B"], "4": ["shows God's part in creating the universe", "shows the existence of a man thousands of years ago", "shows messages inscribed in DNA", "shows a court opinion against Intelligent Design", "shows God's existence"], "5": ["D", "D", "A", "B", "B"], "6": ["A", "C", "B", "A", "D"], "7": ["would-be", "intelligence", "genes", "athletic", "medical", "disclosed", "consideration", "the sum American egg donors expect to be paid", "plus all the costs of medical treatment and insurance", "almost half the cost of fees for the students\u2019 four-year college course."]}, "8": {"1": ["C", "A", "A", "B", "C", "D", "D", "A", "C", "B"], "2": ["D", "A", "A", "B", "A"], "3": ["D", "A", "B", "B", "A"], "4": ["Bill Gates", "Warren Buffet", "KP Singh", "Martha Stewart", "Ronald Lauder", "Luisa Kroll"], "5": ["A", "C", "D", "B", "C"], "6": ["A", "D", "B", "B", "B"], "7": ["greeted", "freshman", "spite", "fluke", "agitated", "faculty", "particularly", "I had the highest average in the freshman class", "Then, she took out a copy of the examination paper", "I was so angry that I started punding her"]}, "9": {"1": ["A", "C", "B", "D", "C", "D", "A", "A", "B", "C"], "2": ["D", "A", "B", "B", "A"], "3": ["D", "B", "B", "C", "C"], "4": ["His brother's two daughters.", "They drove there.", "Children under three.", "They watched the parade.", "He sat down and had a rest for a few minutes."], "5": ["C", "B", "A", "C", "D"], "6": ["B", "A", "D", "C", "B"], "7": ["Authorities", "grant", "opera", "not traditional", "staged", "journalists", "tunnel", "Critics, performers and audience alike were especially excited about the new introduction to the story", "who kills all suitors who cannot answer her three riddles", "Some Chinese critics have complained that the princess has none of the grace of a true Chinese lady"]}, "10": {"1": ["D", "B", "A", "C", "B", "B", "B", "C", "B", "A"], "2": ["D", "B", "C", "A", "B"], "3": ["C", "B", "D", "D", "A"], "4": ["is something not known for sure", "is trying to figure out how dreams help in solving problems", "is something everyone should do", "is important in interpreting dreams", "is something a person might be afraid of"], "5": ["C", "D", "C", "A", "A"], "6": ["A", "D", "C", "B", "A"], "7": ["rarely", "consulting", "renting", "agent", "appointment", "exchange", "praises", "the people who work there actually know where things are and they'll take you right to them", "I will tell you what I really like about Publix", "There's a different grocery store near my house, but all that the employees do there is to grunt"]}};
	
	pattern = new RegExp(".*?book/book25/dj(.*?).php\\?UnitID=(\\d*).*","ig");
	matches = pattern.exec(document.URL);
//	console.log(matches);
	
	if(matches != undefined && matches.length == 3)
		nhce_fill_blank_3(matches[1],matches[2]);
	
	pattern = new RegExp(".*?book/book26/dj(.*?).php\\?UnitID=(\\d*).*","ig");
	matches = pattern.exec(document.URL);
//	console.log(matches);
	
	if(matches != undefined && matches.length == 3)
		nhce_fill_blank_4(matches[1],matches[2]);

	if(document.URL.indexOf("dj24.php") > 0)
		jQuery(".next").attr("href",jQuery(".next").attr("href").replace("dj33.php","dj51.php"));
	/*
	if(document.URL.indexOf("dj53.php") > 0){
		jQuery(".next").attr("href",jQuery(".next").attr("href").replace("dj56.php","dj21.php"));
		jQuery(".next").attr("href",jQuery(".next").attr("href").replace("UnitID="+matches[2],"UnitID="+String(Number(matches[2])+1)));

	}*/
    jQuery(document).keydown(function(){
        if(window.event.keyCode == 78)
            jQuery(".next")[0].click();

    });
    jQuery("body").attr("onkeydown","javascript: if(window.event.keyCode==83)djSubmit();");
	
	
}

function nhce_fill_blank_3(part,unit){
	// Listening - Short Conversation
	if(part == "21"){
		var kid = 1;
		pattern = new RegExp(".*KidID=(\\d*).*","ig");
		matches = pattern.exec(document.URL);
		if(matches != undefined && matches.length == 2)
			kid = Number(matches[1]);
		
		console.log("kid:");
		console.log(kid);
		
		console.log("answer:");
		ans = answer03[unit]["1"][kid-1];
		console.log(ans);
		jQuery("input[value="+ans+"]").attr("checked","checked");
		jQuery("input[value="+ans+"]").parent().parent().css("background-color","yellow");
		
        return;

	}
    
    // Listening - Long Conversation
	if(part == "22"){

		console.log("answer:");
		ans = answer03[unit]["2"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            var start_id = Number(jQuery("input[type=radio]").slice(0,1).attr("name").slice(5));
            radio = jQuery("#Item_"+String(start_id-1+x)+"_"+ans[x-1]);
            radio.attr("checked","checked");
            radio.parent().parent().css("background-color","yellow");
        }
        return;
    }    
    // Listening - Passage
	if(part == "23"){

		console.log("answer:");
		ans = answer03[unit]["3"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            var start_id = Number(jQuery("input[type=radio]").slice(0,1).attr("name").slice(5));
            radio = jQuery("#Item_"+String(start_id-1+x)+"_"+ans[x-1]);
            radio.attr("checked","checked");
            radio.parent().parent().css("background-color","yellow");
        }
        return;
    }  
    
    // Listening - Radio Program
	if(part == "24"){

		console.log("answer:");
		ans = answer03[unit]["4"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            radio = jQuery("input[name=Item_"+String(x-1)+"]");
            radio.val(ans[x-1]);
            radio.css("background-color","yellow");
        }
        return;
    }  
    
    // Homework - Long Conversation
	if(part == "51"){

		console.log("answer:");
		ans = answer03[unit]["5"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            radio = jQuery("#Item_"+String(0+x)+"_"+ans[x-1]);
            radio.attr("checked","checked");
            radio.parent().parent().css("background-color","yellow");
        }
        return;
    }    
    // Homework - Passage
	if(part == "52"){

		console.log("answer:");
		ans = answer03[unit]["6"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            radio = jQuery("#Item_"+String(0+x)+"_"+ans[x-1]);
            radio.attr("checked","checked");
            radio.parent().parent().css("background-color","yellow");
        }
        return;
    }    

    // Homework - Compound Dictation
	if(part == "53"){

		console.log("answer:");
		ans = answer03[unit]["7"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            radio = jQuery("input[name=Item_"+String(x-1)+"]");
            radio.val(ans[x-1]);
            radio.css("background-color","yellow");
        }
        return;
    }  

    


}


function nhce_fill_blank_4(part,unit){
	// Listening - Short Conversation
	if(part == "21"){
		var kid = 1;
		pattern = new RegExp(".*KidID=(\\d*).*","ig");
		matches = pattern.exec(document.URL);
		if(matches != undefined && matches.length == 2)
			kid = Number(matches[1]);
		
		console.log("kid:");
		console.log(kid);
		
		console.log("answer:");
		ans = answer04[unit]["1"][kid-1];
		console.log(ans);
		jQuery("input[value="+ans+"]").attr("checked","checked");
		jQuery("input[value="+ans+"]").parent().parent().css("background-color","yellow");
		
        return;

	}
    
    // Listening - Long Conversation
	if(part == "22"){

		console.log("answer:");
		ans = answer04[unit]["2"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            var start_id = Number(jQuery("input[type=radio]").slice(0,1).attr("name").slice(5));
            radio = jQuery("#Item_"+String(start_id-1+x)+"_"+ans[x-1]);
            radio.attr("checked","checked");
            radio.parent().parent().css("background-color","yellow");
        }
        return;
    }    
    // Listening - Passage
	if(part == "23"){

		console.log("answer:");
		ans = answer04[unit]["3"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            var start_id = Number(jQuery("input[type=radio]").slice(0,1).attr("name").slice(5));
            radio = jQuery("#Item_"+String(start_id-1+x)+"_"+ans[x-1]);
            radio.attr("checked","checked");
            radio.parent().parent().css("background-color","yellow");
        }
        return;
    }  
    
    // Listening - Radio Program
	if(part == "24"){

		console.log("answer:");
		ans = answer04[unit]["4"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            radio = jQuery("input[name=Item_"+String(x-1)+"]");
            radio.val(ans[x-1]);
            radio.css("background-color","yellow");
        }
        return;
    }  
    
    // Homework - Long Conversation
	if(part == "51"){

		console.log("answer:");
		ans = answer04[unit]["5"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            radio = jQuery("#Item_"+String(0+x)+"_"+ans[x-1]);
            radio.attr("checked","checked");
            radio.parent().parent().css("background-color","yellow");
        }
        return;
    }    
    // Homework - Passage
	if(part == "52"){

		console.log("answer:");
		ans = answer04[unit]["6"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            radio = jQuery("#Item_"+String(0+x)+"_"+ans[x-1]);
            radio.attr("checked","checked");
            radio.parent().parent().css("background-color","yellow");
        }
        return;
    }    

    // Homework - Compound Dictation
	if(part == "53"){

		console.log("answer:");
		ans = answer04[unit]["7"];
        console.log(ans);
        for(x = 1; x <= ans.length; x++){
            radio = jQuery("input[name=Item_"+String(x-1)+"]");
            radio.val(ans[x-1]);
            radio.css("background-color","yellow");
        }
        return;
    }  

    


}